### PR TITLE
Concurrent import

### PR DESCRIPTION
**Pull request by kaapstorm**
Jira: [SC-721](https://dimagi-dev.atlassian.net/browse/SC-721)

##### SUMMARY
* Loads blobs concurrently using asyncio. Five workers (an arbitrarily chosen number) pull blobs off a shared queue. Workers load the blobs asynchronously by running `blobdb.copy_blob()` in a spawned thread.
* Loads SQL data concurrently using multi-processing. There is one queue for each database, and one worker for each queue. Each worker runs in its own process.
* No change to loading Couch data. The reward did not seem as great as the other two cases.

Easier to :blowfish: review by commit

The commit "[Don't spawn new processes for each chunk](https://github.com/dimagi/commcare-hq/commit/ff7f47f1062b08308b6ffb3f5ae85649342804bf)" includes a bit of refactoring as well as new code. It might be simplest to start in the `load_objects()` method at ...
```python
    for line in object_strings:
        obj = self.line_to_object(line)
        if obj:
            enqueue_object(dbalias_to_workerqueue, obj)
    terminate_workers(dbalias_to_workerqueue)
    load_stats = collect_stats(dbalias_to_workerqueue)
```
... which will look familiar, and work your way in/out from there.

##### RISK ASSESSMENT / QA PLAN
No need for QA.

Opening as a Draft PR for now, pending testing with a larger set of data.

##### PRODUCT DESCRIPTION
Makes migrations to self-hosted environments faster. Improves the speed of loading data into databases by processing the data in parallel.

### CODE REVIEW

**Comment by snopoke** [comment]
FYI I'm not going to get to this review till sometime next week

File: corehq/apps/dump_reload/sql/load.py [hunk 1]
```diff
@@ -18,6 +22,17 @@
...
+CHUNK_SIZE = 200
+
+
+class DefaultDictWithKey(defaultdict):
+    """
+    A defaultdict that passes ``key`` to its factory
+    """
+    def __missing__(self, key):
+        self[key] = value = self.default_factory(key)
+        return value
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:34]: Could move to bottom of file since this is a utility class only used by code below (newspaper metaphor).

File: corehq/apps/dump_reload/sql/load.py [hunk 2]
```diff
@@ -37,39 +52,55 @@ class SqlDataLoader(DataLoader):
...
+            for __, queue in dbalias_to_workerqueue.values():
+                queue.put(None)
+
+        def collect_stats(dbalias_to_workerqueue):
+            worker_tasks = {wq[0] for wq in dbalias_to_workerqueue.values()}
+            return [t.result() for t in as_completed(worker_tasks)]
+
+        num_aliases = len(settings.DATABASES)
+        manager = mp.Manager()
+        with ProcessPoolExecutor(max_workers=num_aliases) as executor:
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:76]: This is super cool. Way easier than trying to use multiple processes with gevent.

File: corehq/apps/dump_reload/sql/load.py [hunk 3]
```diff
@@ -37,39 +52,55 @@ class SqlDataLoader(DataLoader):
...
             loaded_model_counts.update(model_labels)
+        return object_count, loaded_model_counts
 
-        return sum(total_object_counts), loaded_model_counts
+    def line_to_object(self, line):
+        line = line.strip()
+        if line:
+            obj = json.loads(line)
+            if self.filter_object(obj):
+                return obj
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:103]: nit: explicit is better than implicit
```suggestion
                return obj
        return None
```

File: corehq/apps/dump_reload/sql/load.py [hunk 4]
```diff
@@ -37,39 +52,55 @@ class SqlDataLoader(DataLoader):
...
+        with ProcessPoolExecutor(max_workers=num_aliases) as executor:
+            # Map each db_alias to a queue + a worker task to consume the queue
+            worker_queue_factory = partial(get_worker_queue, executor, manager)
+            # DefaultDictWithKey passes the key to its factory function so that
+            # the worker knows its db_alias without having to figure it out
+            dbalias_to_workerqueue = DefaultDictWithKey(worker_queue_factory)
+
+            for line in object_strings:
+                obj = self.line_to_object(line)
+                if obj:
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:85]: nit
```suggestion
                if obj is not None:
```

File: corehq/apps/dump_reload/sql/load.py [hunk 5]
```diff
@@ -153,35 +200,27 @@ def load_data_for_db(db_alias, objects):
...
+    if hasattr(model, 'partition_attr'):
+        try:
+            partition_value = obj['fields'][model.partition_attr]
+        except KeyError:
+            # in the case of foreign keys the serialized field name is the
+            # name of the foreign key attribute
+            field = [
+                field for field in model._meta.fields
+                if field.column == model.partition_attr
+            ][0]
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:216]: Slightly more defensive. I'm assuming there should always be exactly one.
```suggestion
            field, = [
                field for field in model._meta.fields
                if field.column == model.partition_attr
            ]
```

File: corehq/apps/dump_reload/sql/load.py [hunk 6]
```diff
@@ -37,39 +52,55 @@ class SqlDataLoader(DataLoader):
...
+
+        def terminate_workers(dbalias_to_workerqueue):
+            for __, queue in dbalias_to_workerqueue.values():
+                queue.put(None)
+
+        def collect_stats(dbalias_to_workerqueue):
+            worker_tasks = {wq[0] for wq in dbalias_to_workerqueue.values()}
+            return [t.result() for t in as_completed(worker_tasks)]
+
+        num_aliases = len(settings.DATABASES)
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:74]: Does this thing operate on all databases, or only shard dbs? If only shards, is this intentionally higher than the number of shards? Or maybe it doesn't matter?

**Review by millerdev — COMMENTED** [review]
Looks good to me. I guess there are still tests to fix?

File: corehq/apps/dump_reload/sql/load.py (hunk 2 above)
Comment (kaapstorm) [reply corehq/apps/dump_reload/sql/load.py:76]: I agree. It does get a little fiddly though. You'd think that `multiprocessing.Queue` could communicate between processes, but no, you need to create it using `multiprocessing.Manager` to do that.

File: corehq/apps/dump_reload/sql/load.py (hunk 6 above)
Comment (kaapstorm) [reply corehq/apps/dump_reload/sql/load.py:74]: The original idea was to have a process for each shard, but the code already organises data by db_alias, so it was simpler to assign workers by db_alias instead. Capping the number of workers at "all of the databases" seemed safe. We probably don't have data for all the db_aliases, so we probably won't use all the workers.

Optimally we would want to have one process per core (which is the default value for `max_workers`), and if the number of db_aliases being loaded is greater than the number of cores, run them in different threads across those processes. ... But simple is better than complex.

File: corehq/apps/dump_reload/sql/load.py [hunk 7]
```diff
@@ -58,6 +54,8 @@ def terminate_workers(dbalias_to_workerqueue):
 
         num_aliases = len(settings.DATABASES)
         manager = mp.Manager()
+        Executor = ProcessPoolExecutor if os.cpu_count() >= num_aliases \
+            else ThreadPoolExecutor
```
Comment (millerdev) [inline corehq/apps/dump_reload/sql/load.py:58]: (How) do you know if it is better to use `ThreadPoolExecutor` with low CPU count? I would expect processes would always either be better or worse, depending on the type of task being done.

Comment (kaapstorm) [reply corehq/apps/dump_reload/sql/load.py:58]: There is more than one factor to consider, but the general heuristic I've been applying:

* Task uses a lot of CPU: Use processes
* Task waits for other services: Use threads

Database tasks are a bit of both. I watched my CPU load as I tested, and the processes didn't max out the cores, but they spread noticeable load across them, so I leaned to using processes rather than threads.

As you know, processes are assigned to cores. So if the number of processes is greater than the number of cores, you've got cores sharing processes, which is kind of like using very expensive threads. If you're running this code on a bare-metal system with 8 cores and 5 DB aliases, processes are perfect. But if you're on a VM with a single core and 9 DB aliases, threads will be better.

It could be that the overhead for processes is not a huge consideration because this is a long-running task on machines that ought to have the RAM, but the docs in the library persuaded me that I was thinking along the right lines:

From Python 3.8 ProcessPoolExecutor docstring:

> max_workers: ... If None or not given then as many
> worker processes will be created as the machine has processors.

From Python 3.8 ThreadPoolExecutor comments:

>         if max_workers is None:
>             # We use cpu_count + 4 ...
>             # But we limit it to 32 to avoid consuming surprisingly
>             # large resource on many core machine.

(Prior to 3.8, ProcessPoolExecutor and ThreadPoolExecutor both used "cpu_count + 4". The fact that ProcessPoolExecuter changed, in my mind, added weight to my belief.)

But I could definitely be wrong about this, so if you think that the nature of the task outweighs the hardware its running on, I'd be happy to switch to a single approach. Also, if you think I should lean towards threads instead of processes, I'm also open to being persuaded.

Comment (millerdev) [reply corehq/apps/dump_reload/sql/load.py:58]: > if you're on a VM with a single core and 9 DB aliases, threads will be better.

To me this seems not necessarily a safe assumption, but it probably depends a lot on the workload. Threads are limited by the GIL, and therefore task switches are managed by the Python runtime. I've heard rumors that this can be inefficient in some cases (watch the first 4 minutes of [this talk](https://pyvideo.org/pycon-us-2010/pycon-2010--understanding-the-python-gil---82.html) for a demonstration of that). Maybe this has changed in modern Python versions?

Processes on the other hand are managed by the OS, which I assume has more robust not to mention far more stress-tested task switching mechanics than the Python runtime.

I admit what I'm saying here is pretty hand wavy, so if your experience indicates that what you're doing is good then feel free to ignore me.

Comment (kaapstorm) [reply corehq/apps/dump_reload/sql/load.py:58]: Oh man, those first four minutes are awesome. I will watch the rest later.

Yes, I know a little bit about the GIL. And you're right, it does get more complicated, because, if I remember correctly, `threading` threads are also created using the OS, but `asyncio` threads are not? (Maybe? I'm not sure. Maybe Py3 no longer uses the OS for any threads. I'd google it, but my electricity is going to get cut soon. But definitely worth exploring a bit more.)

Comment (millerdev) [reply corehq/apps/dump_reload/sql/load.py:58]: I should also add, after watching the entire talk, that near the end he shows a slide that says

> Don't use this talk to justify not using threads

Which is pretty much exactly what I did here. So I abused that talk, sorry for that. Also, there are hints in the talk that the situation has probably gotten much better in Python 3 (since Python 3.2 at least). But I think it still makes sense to measure the performance of your program and verify that it meets expectations.

Comment (kaapstorm) [reply corehq/apps/dump_reload/sql/load.py]: So, after managing to see how Travis is running TestSQLDumpLoadShardedModels in test_sql_dump_load.py, I discovered the issue: 

When Travis uses processes, the test passes. But when Travis uses threads, and adds a domain_migration_flags.domainmigrationprogress record to the "default" DB, the model's `save()` call hangs. It does not throw an exception.

I haven't dug any deeper. My first guess is that it's related to locking, but I'm not sure why the problem doesn't happen when using processes.

But maybe I should accept this as the "performance measure" I'm looking for :upside_down_face: and come back to the "why" some other time.

**Review by millerdev — APPROVED** [review]
